### PR TITLE
8321688: Build on linux with GCC 7.5.0 fails after 8319577

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

This patch fixes the building failure introduced by [JDK-8319577](https://bugs.openjdk.org/browse/JDK-8319577) in old GCC version (linux & GCC 7.5.0 locally).

Thanks for the review.

 Best Regards,
-- Guoxiong

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8321688](https://bugs.openjdk.org/browse/JDK-8321688): Build on linux with GCC 7.5.0 fails after 8319577 (**Bug** - P4)

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17047/head:pull/17047` \
`$ git checkout pull/17047`

Update a local copy of the PR: \
`$ git checkout pull/17047` \
`$ git pull https://git.openjdk.org/jdk.git pull/17047/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17047`

View PR using the GUI difftool: \
`$ git pr show -t 17047`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17047.diff">https://git.openjdk.org/jdk/pull/17047.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17047#issuecomment-1849261465)

### CODE REVIEW

**Comment by magicus** [comment]
/label -build

Just because a code change cause a compilation error, it does not mean that it is relevant for the build group.

**Review by kimbarrett — COMMENTED** [review]
Please hold off on this change while I follow up on JDK-8319577.  This code
requires C++17, which was turned on for a relevant small subset of the JDK by
that change.  There was some discussion there about version limiting that
change, but that doesn't seem to have been done, with this breakage being a
consequence.

**Comment by magicus** [comment]
@kimbarrett There were a long discussion about the C++17 support. The end result, which led me to approve the build changes, was the claims made beginning with https://github.com/openjdk/jdk/pull/16534#discussion_r1414570644:

>  [...] the updated code which added guards to check for GCC version >= 7.5 in src/java.base/linux/native/libsimdsort/{avx2-linux-qsort.cpp, avx512-linux-qsort.cpp}. GCC >= 7.5 is needed to compile libsimdsort using C++17 features. Made sure that OpenJDK builds without errors using both GCC 7.5 and GCC 6.4.

> The GCC complier for versions 6 (and even 5) silently ignores the flag -std=c++17

Somehow the JDK compiled with 7.5 for the author of that patch, but fails for the author of this patch. I don't understand how this both can be true.

**Comment by lgxbslgx** [comment]
> Somehow the JDK compiled with 7.5 for the author of that patch, but fails for the author of this patch. I don't understand how this both can be true.

I noticed my previous build is `slowdebug`. Today, I tested the `release` and `fastdebug` locally. These builds succeeded. 
So only the `slowdebug` build fails. What's the difference between `fastdebug` and `slowdebug`?
It seems that `slow-debug` disables some optimizations, such as constant propagation, which may cause this issue.
Should we adjust the concrete code (like this PR) or revise the build system to avoid the potential problem in the future?

**Comment by kimbarrett** [comment]
The files being proposed for change are cloned copies of a 3rd-party library.
We may want to take updates from that library in the future. Having local
modifications makes that more difficult.

Also, this library uses C++17.  GCC support for C++17 was experimental until
GCC 9.  As such, it shouldn't be entirely surprising if an earlier version of
GCC might run into problems with it.

We've been discussing switching to using C++17 JDK-wide for some time, and I
think that may happen soon.  At that point the minimum supported version of
GCC will be changed to at least GCC 9.

So I think the answer here is that building this lbrary with GCC 7.5 is not
supported and a newer version should be used. So the change should be here:
https://github.com/openjdk/jdk/blob/b31454e36234091c3827c3b4d07f62345cb0cee4/src/java.base/linux/native/libsimdsort/simdsort-support.hpp#L34-L37
to require a later GCC version. That way, if building with an older version of
GCC then the code in the library won't be built and the library won't be used
by the JDK (because the relevant symbols won't be found).

But GCC 7.5 will probably not be supported for building the JDK for much longer.

**Comment by lgxbslgx** [comment]
@kimbarrett Thanks for the guide. I updated the code just now. All the `release`, `fastdebug` and `slowdebug` builds succeeded locally with GCC7.5.

File: src/java.base/linux/native/libsimdsort/simdsort-support.hpp [hunk 1]
```diff
@@ -31,8 +31,8 @@
 #define assert(cond, msg) { if (!(cond)) { fprintf(stderr, "assert fails %s %d: %s\n", __FILE__, __LINE__, msg); abort(); }}
 
 
-// GCC >= 7.5 is needed to build AVX2 portions of libsimdsort using C++17 features
-#if defined(_LP64) && (defined(__GNUC__) && ((__GNUC__ > 7) || ((__GNUC__ == 7) && (__GNUC_MINOR__ >= 5))))
+// GCC >= 9.1 is needed to build AVX2 portions of libsimdsort using C++17 features
+#if defined(_LP64) && (defined(__GNUC__) && ((__GNUC__ > 9) || ((__GNUC__ == 9) && (__GNUC_MINOR__ >= 1))))
```
Comment (kimbarrett) [inline src/java.base/linux/native/libsimdsort/simdsort-support.hpp:35]: Have you tested with gcc 9?  Or is this just supposition based on gcc9 having removed the experimental
status for C++17?

**Comment by sviswa7** [comment]
Let us restrict the libsimdsort.so build to GCC 8.4.0 with the following change:
`diff --git a/src/java.base/linux/native/libsimdsort/simdsort-support.hpp b/src/java.base/linux/native/libsimdsort/simdsort-support.hpp
index c73fd7920d8..2a2946cff82 100644
--- a/src/java.base/linux/native/libsimdsort/simdsort-support.hpp
+++ b/src/java.base/linux/native/libsimdsort/simdsort-support.hpp
@@ -31,9 +31,9 @@
 #define assert(cond, msg) { if (!(cond)) { fprintf(stderr, "assert fails %s %d: %s\n", __FILE__, __LINE__, msg); abort(); }}


-// GCC >= 7.5 is needed to build AVX2 portions of libsimdsort using C++17 features
-#if defined(_LP64) && (defined(__GNUC__) && ((__GNUC__ > 7) || ((__GNUC__ == 7) && (__GNUC_MINOR__ >= 5))))
+// GCC >= 8.4 is needed to build AVX2 portions of libsimdsort using C++17 features
+#if defined(_LP64) && (defined(__GNUC__) && ((__GNUC__ > 8) || ((__GNUC__ == 8) && (__GNUC_MINOR__ >= 4))))
 #define __SIMDSORT_SUPPORTED_LINUX
 #endif

-#endif //SIMDSORT_SUPPORT_HPP
\ No newline at end of file
+#endif //SIMDSORT_SUPPORT_HPP`
I have verified that with the above change the builds (release, fastdebug, slowdebug) all succeed with GCC 8.4.0 as well as prior GCC like GCC7.5.0 and the test/jdk/java/util/Arrays/Sorting.java passes successfully with these builds.

**Comment by lgxbslgx** [comment]
> Have you tested with gcc 9? Or is this just supposition based on gcc9 having removed the experimental
status for C++17?

I have not tested GCC 8 and 9. @sviswa7 seems to test them.

> I have verified that with the above change the builds (release, fastdebug, slowdebug) all succeed with GCC 8.4.0 as well as prior GCC like GCC7.5.0 and the test/jdk/java/util/Arrays/Sorting.java passes successfully with these builds.

Thanks for your tests. But from the description of the [GCC document](https://gcc.gnu.org/projects/cxx-status.html), shown below, it may be good to skip GCC 8 and use GCC 9 directly if we want to switch to C++17.

>  Some C++17 features are available since GCC 5, but support was experimental and the ABI of C++17 features was not stable until GCC 9.

What do you think about it?

**Comment by sviswa7** [comment]
@lgxbslgx  We would like to keep GCC 8.4.0 as the minimum.

**Comment by sviswa7** [comment]
@kimbarrett I meant to say that since the libsimdsort works with GCC 8.4.0, the #define guard in libsimdsort sources could be restricted to just that and we don't have to artificially raise it to GCC 9. Do you think that is an issue?

**Comment by kimbarrett** [comment]
> @kimbarrett I meant to say that since the libsimdsort works with GCC 8.4.0, the #define guard in libsimdsort sources could be restricted to just that and we don't have to artificially raise it to GCC 9. Do you think that is an issue?

I don't think we should be depending on an experimental compiler feature when there are alternatives.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Review by sviswa7 — APPROVED** [review]

**Comment by lgxbslgx** [comment]
@kimbarrett @sviswa7 Thanks for your reviews. Integrating.

/integrate

**Pull request closed without merging**